README.md
# Twinkle

Binding of [Twinkle](https://github.com/kirisaki/twinkle) for Rust.

[![Actions Status](https://github.com/kirisaki/twinkle-rust/workflows/test-twinkle/badge.svg)](https://github.com/kirisaki/twinkle-rust/actions)
[![twinkle at crates.io](https://img.shields.io/crates/v/twinkle.svg)](https://crates.io/crates/twinkle)
[![twinkle at docs.rs](https://docs.rs/twinkle/badge.svg)](https://docs.rs/crate-name)

## Usage

Cargo.toml

```toml
[dependencies]
twinkle = "0.1"
futures = "0.3"
tokio = { version = "0.2", features = ["full"] }
```

main.rs

```rust
use futures::future::{join};

#[tokio::main]
async fn main(){
    let (client, manager) = twinkle::open("127.0.0.1:3000").await.unwrap();
    join(manager.run(), your_app(client));
}

async fn your_app(mut c: twinkle::Client) {
    c.ping().await;
    c.set(b"hoge".to_vec(), b"foo".to_vec()).await;
    c.get(b"hoge".to_vec()).await;
    c.unset(b"hoge".to_vec()).await;
}

```

## License

[BSD-3-Clause](https://opensource.org/licenses/BSD-3-Clause)


CHANGELOG.md
# Change Log


## [0.1.3] - 2019-12-26

### Feature

- Bundle `Listener` and `Dispatcher` into `Manager`.


## [0.1.2] - 2019-12-21

### Feature

- Operations are resended with heuristic way.

## [0.1.0] - 2019-12-14

### Feature

- Basic functions.


src/client.rs

use std::sync::Arc;
use std::collections::HashMap;
use std::io::{Error, ErrorKind};

use futures::future::join;

use tokio::net::{UdpSocket, ToSocketAddrs};
use tokio::sync::Mutex;
use tokio::sync::mpsc::{Receiver, Sender, channel};
use tokio::time::{delay_for, Duration};

use crate::dispatcher::*;
use crate::listener::*;
use crate::request::*;
use crate::types::*;

/// `Client` has `Sender<Bytes>` for sending bytes
/// and `Arc<Mutex<Table>>` which has status of sended packets.
#[derive(Clone)]
pub struct Client {
    pub chan: Sender<Bytes>,
    pub tabl: Arc<Mutex<Table>>,
}

impl Client {
    /// Open a connection to a twinkle server.
    pub async fn open<A: ToSocketAddrs>(addr: A) -> Result<(Client, Manager), std::io::Error> {
        let sock = UdpSocket::bind("0.0.0.0:0").await?;
        sock.connect(addr).await?;
        let (txc, rxc) = channel(1024*1024*1024);

        let tabl = Arc::new(Mutex::new(HashMap::new()));
        let c = Client{chan: txc, tabl: tabl.clone()};
        let m = Manager::new(rxc, sock, tabl);
        Ok((c, m))
    }

    /// Send `Ping`.
    pub async fn ping(&mut self) -> Result<Bytes, std::io::Error> {
        send(self, Request::Ping).await
    }
    /// Send `Get`.
    pub async fn get(&mut self, key: Bytes) -> Result<Bytes, std::io::Error> {
        send(self, Request::Get(key)).await
    }
    /// Send `Set`.
    pub async fn set(&mut self, key: Bytes, val: Bytes) -> Result<Bytes, std::io::Error> {
        send(self, Request::Set(key, val)).await
    }
    /// Send `Unset`.
    pub async fn unset(&mut self, key: Bytes) -> Result<Bytes, std::io::Error> {
        send(self, Request::Unset(key)).await
    }
}

async fn send(client: &mut Client, request: Request) -> Result<Bytes, std::io::Error>{
    let packet = request.issue()?;
    for _ in 0..10 {
        let mut rx = packet.clone().dispatch(client).await?;
        for i in 0..5u32 {
            delay_for(Duration::from_millis(2u64.pow(i))).await;
            match rx.try_recv() {
                Ok(v) => return v,
                Err(_) => {},
            };

        };
    }
    Err(Error::new(ErrorKind::Other, "request timeout"))
}

/// `Manager` manages to send packets and to receive packets.
pub struct Manager {
    dispatcher: Dispatcher,
    listener: Listener,
}

impl Manager {
    fn new(chan: Receiver<Bytes>, sock: UdpSocket, tabl: Arc<Mutex<Table>>) -> Manager{
        let (rxs, txs) = sock.split();
        Manager{
            dispatcher: Dispatcher{chan, sock: txs},
            listener: Listener{sock: rxs, tabl}
        }
    }
    /// Run the `Manager`.
    pub async fn run(mut self) -> Result<(), std::io::Error>{
        let _ = join(
            self.dispatcher.run(),
            self.listener.run()
        ).await;
        Ok(())
    }
}


src/request.rs
use uuid::Uuid;

use crate::packet::*;
use crate::types::*;
use crate::errors::*;

#[derive(Debug, PartialEq)]
pub enum Request {
    Ping,
    Get(Bytes),
    Set(Bytes, Bytes),
    Unset(Bytes),
}

impl Request {
    pub fn issue(self) -> Result<(Packet), std::io::Error> {
        let uuid = Uuid::new_v4();
        let cont = match self {
            Request::Ping => {
                let mut c = vec![0x01];
                c.append(&mut uuid.as_bytes().to_vec());
                c
            },
            Request::Get(key) => {
                let ks = usize_to_bytes(key.len())?;
                let mut c = vec![0x02];
                c.append(&mut uuid.as_bytes().to_vec());
                c.append(&mut ks.to_vec());
                c.append(&mut key.to_vec());
                c
            },
            Request::Set(key, val) => {
                let ks = usize_to_bytes(key.len())?;
                let mut c = vec![0x03];
                c.append(&mut uuid.as_bytes().to_vec());
                c.append(&mut ks.to_vec());
                c.append(&mut key.to_vec());
                c.append(&mut val.to_vec());
                c
            },
            Request::Unset(key) => {
                let ks = usize_to_bytes(key.len())?;
                let mut c = vec![0x04];
                c.append(&mut uuid.as_bytes().to_vec());
                c.append(&mut ks.to_vec());
                c.append(&mut key.to_vec());
                c
            },
        };
        Ok(Packet{cont, uuid})
    }
}

fn usize_to_bytes(u: usize) -> Result<[u8; 2], std::io::Error> {
    if u > From::from(u16::max_value()){
        Err(std::io::Error::new(std::io::ErrorKind::Other, "keylen too long"))
    } else {
        let ks = u.to_be_bytes();
        Ok([ks[6], ks[7]])
    }
}


src/listener.rs
use std::io::{Error, ErrorKind};
use std::sync::Arc;

use uuid::Uuid;

use tokio::sync::Mutex;
use tokio::net::udp::RecvHalf;

use crate::types::*;
use crate::errors::*;

pub struct Listener {
    pub sock: RecvHalf,
    pub tabl: Arc<Mutex<Table>>,
}

impl Listener {
    pub async fn run(&mut self) -> Result<(), std::io::Error> {
        let Listener{sock, tabl} = self;
        let mut buf = vec![0; BUF_SIZE];
        loop {
            let amt = sock.recv(&mut buf).await?;
            let e = std::io::Error::new(std::io::ErrorKind::Other, "failed parsing");
            if amt < UUID_LEN + 1 {
                return Err(e);
            } else {
                let uuid = match Uuid::from_slice(&buf[1..UUID_LEN+1]) {
                    Ok(v) => v,
                    Err(e) => return Err(std::io::Error::new(std::io::ErrorKind::Other, e.to_string())),
                };
                let mut tabl = tabl.lock().await;
                let val = if UUID_LEN +1 == amt {
                    vec![]
                } else {
                    buf[UUID_LEN + 2..amt].to_vec()
                };
                match tabl.get_mut(&uuid) {
                    Some(c) => {
                        if buf[0] == 0x01 {
                            let _ = c.send(Ok(val)).await;
                        } else {
                            let _ = c.send(Err(Error::new(ErrorKind::Other, "command failed"))).await;
                        };
                    },
                    None => return Err(e),
                };
                tabl.remove(&uuid);
            }

        }
    }
}


src/lib.rs
//! Client library for [twinkle](https://github.com/kirisaki/twinkle).
//!
//! Cargo.toml
//!
//! ```toml
//! [dependencies]
//! twinkle = "0.1"
//! futures = "0.3"
//! tokio = { version = "0.2", features = ["full"] }
//! ```
//! 
//! main.rs
//! 
//! ```rust
//! use futures::future::{join};
//! use twinkle::client::Client;
//! 
//! #[tokio::main]
//! async fn main(){
//!     let (client, manager) = Client::open("127.0.0.1:3000").await.unwrap();
//!     join(manager.run(), your_app(client));
//! }
//! 
//! async fn your_app(mut c: Client) {
//!     c.ping().await;
//!     c.set(b"hoge".to_vec(), b"foo".to_vec()).await;
//!     c.get(b"hoge".to_vec()).await;
//!     c.unset(b"hoge".to_vec()).await;
//! }
//! 
//! ```

pub mod client;
mod listener;
pub mod request;
mod packet;
mod dispatcher;
mod types;
mod errors;

use futures::future::join;
use tokio::time::timeout;
use std::time::{Duration};
use std::io::{Error, ErrorKind};


#[cfg(test)]
mod tests {
    use crate::*;
    use crate::client::Client;
    const WAIT: Duration = std::time::Duration::from_secs(1);

    #[tokio::test]
    async fn test_ping() -> Result<(), std::io::Error> {
        let (mut client, manager) = Client::open("127.0.0.1:3000").await?;
        let (_, res) = join(
            timeout(WAIT, manager.run()),
            client.ping()).await;
        match res {
            Ok(_) => Ok(()),
            Err(e) => Err(e),
        }
    }

    #[tokio::test]
    async fn test_get_not_found() -> Result<(), std::io::Error> {
        let (mut client, manager) = Client::open("127.0.0.1:3000").await?;
        let (_, res) = join(
            timeout(WAIT, manager.run()),
            timeout(WAIT, async move {
                client.unset(b"hoge".to_vec()).await;
                client.get(b"hoge".to_vec()).await
        })).await;
        match res.unwrap() {
            Ok(_) => Err(Error::new(ErrorKind::Other, "found key")),
            Err(_) => Ok(()),
        }
    }

    #[tokio::test]
    async fn test_set_and_get() {
        let (mut client, manager) = Client::open("127.0.0.1:3000").await.unwrap();
        let (_, res) = join(
            timeout(WAIT, manager.run()),
            timeout(WAIT, async move {
                client.set(b"fuga".to_vec(), b"foo".to_vec()).await;
                client.get(b"fuga".to_vec()).await.unwrap()
        })).await;
        assert_eq!(res.unwrap(), b"foo".to_vec());
    }
}
